### DISCUSSION TITLE
`jose-elias-alvarez/typescript.nvim` always loaded when using `extras.lang.typescript`

### DISCUSSION THREAD

**Discussion by loiccoyle** [LazyVim/LazyVim#411] [Q&A]
I was playing around with the `typescript` extra, and I noticed the `jose-elias-alvarez/typescript.nvim` plugin is loaded regardless of the filetype, it is loaded as soon as `lspconfig` or `null-ls` are loaded. While this isn't a big problem as it is just one plugin, but as more `extras` get added for different languages it might become one if they have the same behaviour. 

The same behaviour is present for the (now closed for some reason) `rust-tools` extras PR https://github.com/LazyVim/LazyVim/pull/24, `rust-tools` is loaded with `lspconfig`, as `extras.lang.rust` follows the same "template" as `extras.lang.typescript`.

What would be the best way to include these language-specific lsp extension plugins  (`typescript.nvim`, `rust-tools.nvim`, ....) while only loading them when needed?

I think, in an ideal world, these plugins should only be loaded when their relevant lsp servers are running.

**Comment by folke** [comment]
That's just how lspconfig works. lspconfig needs to be setup so it can detect when it needs to load. For typescript, it's typescript.nvim that sets up lspconfig, so this has to be done at startup

**Reply by loiccoyle** [reply]
I think I got it to work.
```lua
return {
  -- add typescript to treesitter
  {
    "nvim-treesitter/nvim-treesitter",
    opts = function(_, opts)
      if type(opts.ensure_installed) == "table" then
        vim.list_extend(opts.ensure_installed, { "typescript", "tsx" })
      end
    end,
  },
  -- Setup typescripts.nvim to lazy load when in a typescript file.
  -- We don't add it as an lspconfig dependency as it will always get loaded with lspconfig.
  -- The lspconfig server settings are added to the opts and used to setup typescript.
  {
    "jose-elias-alvarez/typescript.nvim",
    ft = { "typescript", "typescriptreact" },
    dependencies = {
      {
        "jose-elias-alvarez/null-ls.nvim",
        opts = function(_, opts)
          opts.sources = vim.list_extend(opts.sources, { require("typescript.extensions.null-ls.code-actions") })
        end,
      },
    },
    config = function(_, opts)
      -- pull in the lspconfig server settings and use them for typescript setup
      local lsp_opts = require("lazyvim.util").opts("nvim-lspconfig")
      opts.server = lsp_opts.servers.tsserver
      require("typescript").setup(opts)
    end,
  },
  -- correctly setup lspconfig
  {
    "neovim/nvim-lspconfig",
    opts = {
      -- make sure mason installs the server
      servers = {
        tsserver = {
          settings = {
            completions = {
              completeFunctionCalls = true,
            },
          },
        },
      },
      setup = {
        tsserver = function()
          require("lazyvim.util").on_attach(function(client, buffer)
            if client.name == "tsserver" then
              -- stylua: ignore
              vim.keymap.set("n", "<leader>co", function() require("typescript").actions.organizeImport() end, { buffer = buffer, desc = "Organize Imports" })
              -- stylua: ignore
              vim.keymap.set("n", "<leader>cR", "<cmd>TypescriptRenameFile<CR>", { desc = "Rename File", buffer = buffer })
            end
          end)
          return true
        end,
      },
    },
  },
}
```
I don't actually use `typescript` so I haven't been able to test it thoroughly, but as far as I can tell `typescript.nvim` is only loaded in `ts` files and the lsp server is correctly setup. If you want I can open a PR.

**Reply by loiccoyle** [reply]
Hey @folke, can get your opinion on this?

**Reply by folke** [reply]
I already did. It works fine as is. Won't change it since that could lead to issues.
